Add FSNER example in research_projects (#13155)
huggingface/transformers: examples/research_projects/fsner/README.md

# FSNER

Implemented by [sayef](https://huggingface.co/sayef).

## Overview

The FSNER model was proposed in [Example-Based Named Entity Recognition](https://arxiv.org/abs/2008.10570) by Morteza Ziyadi, Yuting Sun, Abhishek Goswami, Jade Huang, Weizhu Chen. To identify entity spans in a new domain, it uses a train-free few-shot learning approach inspired by question-answering.

## Abstract
----
> We present a novel approach to named entity recognition (NER) in the presence of scarce data that we call example-based NER. Our train-free few-shot learning approach takes inspiration from question-answering to identify entity spans in a new and unseen domain. In comparison with the current state-of-the-art, the proposed method performs significantly better, especially when using a low number of support examples.

## Model Training Details
-----

| identifier        | epochs           | datasets  |
| ---------- |:----------:| :-----:|

> | sayef/fsner-bert-base-uncased      | 10 | ontonotes5, conll2003, wnut2017, and fin (Alvarado et al.). |

Reviewer: Perhaps you can make sayef/fsner-bert-base-uncased a hyperlink to the model on the hub

Author: I have added a hyperlink to the model on the hub.

## Installation and Example Usage
------

> You can use the FSNER model in two ways-

Reviewer: ```suggestion
You can use the FSNER model in two ways:
```

> 1. Install as a package: `python setup.py install` and import as following code example -

Reviewer: ```suggestion
1. Install as a package: `python setup.py install` and import the model as shown in the code example below
```

    or

> 2. Change directory to `src` and import as following code example -

Reviewer: ```suggestion
2. Change directory to `src` and import the model as shown in the code example below
```

```python
from fsner import FSNERModel, FSNERTokenizerWrapper

model = FSNERModel("sayef/fsner-bert-base-uncased")

> tokenizer = FSNERTokenizerUtils("sayef/fsner-bert-base-uncased")

Reviewer: I see you're importing `FSNERTokenizerWrapper`, but then using `FSNERTokenizerUtils`.

Author: Mistake. Fixed in my latest [commit](https://github.com/huggingface/transformers/pull/13155/commits/bffa29558ed181a057a04db74f426f615219b4c5).

> # size of query and supports must be same. If you want to find all the entitites in one particular query, just repeat same query n times where n is the size of supports (or entities).

Reviewer: ```suggestion
# size of query and supports must be the same. If you want to find all the entitites in one particular query, just repeat the same query n times where n is equal to the number of supports (or entities).
```

query = [
    'KWE 4000 can reach with a maximum speed from up to 450 P/min an accuracy from 50 mg',
    'I would like to order a computer from eBay.',
]

# each list in supports are the examples of one entity type

> 

Reviewer: ```suggestion
```

# wrap entities around with [E] and [/E] in the examples

> 

Reviewer: ```suggestion
```

supports = [
        [
           'Horizontal flow wrapper [E] Pack 403 [/E] features the new retrofit-kit „paper-ON-form“',
           '[E] Paloma Pick-and-Place-Roboter [/E] arranges the bakery products for the downstream tray-forming equipment',
           'Finally, the new [E] Kliklok ACE [/E] carton former forms cartons and trays without the use of glue',
           'We set up our pilot plant with the right [E] FibreForm® [/E] configuration to make prototypes for your marketing tests and package validation',
           'The [E] CAR-T5 [/E] is a reliable, purely mechanically driven cartoning machine for versatile application fields'
        ],
        [
            "[E] Walmart [/E] is a leading e-commerce company",
            "I recently ordered a book from [E] Amazon [/E]",
            "I ordered this from [E] ShopClues [/E]",
            "Fridge can be ordered in [E] Amazon [/E]",
            "[E] Flipkart [/E] started it's journey from zero"
        ]
   ]

> 

Reviewer: ```suggestion
```
